cpp/examples/arrow/engine_substrait_example.cc

```diff
@@ -0,0 +1,128 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations
+// under the License.
```

Comment:
Should this be made a cookbook example?

Reply:
We definitely can. I was planning on creating a more comprehensive one for the cookbook. 
With a few filters and functions. Waiting for https://github.com/apache/arrow/pull/12664 this to be finalised. 
What do you think?

Reply:
yes, let's do that. I created an issue here:

https://github.com/apache/arrow-cookbook/issues/168
https://github.com/apache/arrow-cookbook/issues/169

Reply:
I can write a detailed example than what we have presented here in the cookbook. Or remove it from this PR and add it to cookbook later on. What do you think?

Reply:
I'm not sure the example adds much to this PR, we can save it for the cookbook? (We can move the code into the issue comments to make sure we have it around or something.)